harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,440 @@
+import abc
+import datetime
+import logging
+import queue
+import threading
+import time
+import typing
+
+import psutil
+
+from determined import core
+from determined.common import api, constants
+from determined.common.api import bindings
+
+try:
+    import pynvml
+except ImportError:
+    pynvml = None
+
+logger = logging.getLogger("determined.core")
+
+
+class ProfilerContext:
+    """
+    ``ProfilerContext`` gives access to the system profiling feature within Determined.
+
+    It is responsible for collecting system metrics at specified time intervals and reporting
+    them to the master.
+
+    This class is available through the ``core.Context`` and can be turned on/off as such:
+
+    .. code::
+
+       core_context.profiler.on()
+       core_context.profiler.off()
+
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        distributed: core.DistributedContext,
+    ) -> None:
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+        self._distributed = distributed
+        self._on = False
+
+        self._collector: typing.Optional[_Collector] = None
+        self._shipper: typing.Optional[_Shipper] = None
+
+    def on(self, interval: int = 1) -> None:
+        """
+        Turns system profiling functionality on.
+
+        This method creates two threads, one that collects system metrics at specified time
+        intervals, and another that ships them to the master.
+
+        .. note::
+
+            This method is idempotent: if profiling is already on, this method is effectively a
+            no-op.
+
+        Arguments:
+            interval: time (in seconds) between each metric collection.
+
+        """
+        if self._on:
+            return
+
+        # Only enable on local chief workers.
```

Comment:
In the event that this code stays, this seems important enough to put into a docstring.

Also interested in why `local_rank` instead of `cross_rank` or `rank` (or even `is_chief` or `is_local_chief`)